geode-core/src/integrationTest/java/org/apache/geode/internal/datasource/GemFireTransactionDataSourceJUnitTest.java

```diff
@@ -0,0 +1,151 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+/*
+ * GemFireTransactionDataSourceJUnitTest.java JUnit based test
+ *
+ * Created on May 2, 2019, 1:24 PM
+ */
+package org.apache.geode.internal.datasource;
+
+import static org.apache.geode.distributed.ConfigurationProperties.CACHE_XML_FILE;
+import static org.apache.geode.distributed.ConfigurationProperties.MCAST_PORT;
+import static org.apache.geode.test.awaitility.GeodeAwaitility.await;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.sql.SQLException;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+
+import javax.naming.Context;
+import javax.sql.XAConnection;
+import javax.sql.XADataSource;
+import javax.transaction.SystemException;
+import javax.transaction.TransactionManager;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import org.apache.geode.cache.Cache;
+import org.apache.geode.cache.CacheFactory;
+import org.apache.geode.distributed.DistributedSystem;
+import org.apache.geode.util.test.TestUtil;
+
+public class GemFireTransactionDataSourceJUnitTest {
+
+  private TransactionManager transactionManager;
+  private XAConnection xaconn;
+  private XADataSource xads;
+
+  private static Properties props = null;
+  private static DistributedSystem ds1 = null;
+  private static Cache cache = null;
+
+  @Before
+  public void setUp() throws Exception {
+    props = new Properties();
+    props.setProperty(MCAST_PORT, "0");
+    String path =
+        TestUtil.getResourcePath(GemFireTransactionDataSourceJUnitTest.class, "/jta/cachejta.xml");
+    props.setProperty(CACHE_XML_FILE, path);
+    ds1 = DistributedSystem.connect(props);
+    cache = CacheFactory.create(ds1);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    ds1.disconnect();
+  }
+
+  @Test
+  public void testExceptionHandlingRegisterTranxConnection() throws Exception {
+    boolean exceptionoccurred = false;
+    Context ctx = cache.getJNDIContext();
+    GemFireTransactionDataSource ds =
+        (GemFireTransactionDataSource) ctx.lookup("java:/XAPooledDataSource");
+
+    GemFireConnectionPoolManager gcpm = (GemFireConnectionPoolManager) ds.getConnectionProvider();
+    TranxPoolCacheImpl tpci = (TranxPoolCacheImpl) gcpm.getConnectionPoolCache();
+    ds.getConnection();
+    // get connection to activate clean thread, which will sleep for 20 seconds
+    // also duration of connection is cca. 20 secs
+
+    // wait for 10 secs (half of clean thread interval)
+    await().pollDelay(10, TimeUnit.SECONDS).atMost(15, TimeUnit.SECONDS)
```

Comment:
I recommend leaving off the `pollDelay` and `atMost`. We standardized on using the default timeout in GeodeAwaitility to avoid spurious failures on super slow machines, systems with super busy CPU, or if the JVM hits a stop-the-world garbage collection (which can typically halt things for 2 minutes).

Reply:
Thanks for comments. Do you have any proposal how to delay some actions for 10 secs.

Reply:
> await().untilAsserted(() -> assertEquals(1, tpci.getActiveCacheSize()));

No need to delay this assertion for 10 seconds. `untilAsserted` will continue retrying until either the assertion passes or it times out (5 minutes).

Reply:
Maybe this part is misleading, but for my test I need delay, and this is why I put pollDelay (because assertion is fullfield).

Reply:
Testcase was design, to fail in case this solution is not applied. But if it is not needed, I can delete sleep.